contract/src/utils.rs

```diff
@@ -390,3 +365,336 @@ pub(crate) fn is_token_burned(token_identifier: &TokenIdentifier) -> bool {
+            storage::new_dictionary(&dictionary_name)
+                .unwrap_or_revert_with(NFTCoreError::FailedToCreateDictionary);
+        }
+        max_number_of_pages
+    }
+}
+
+pub(crate) fn update_forward_and_reverse_token_trackers(
+    current_number_of_minted_tokens: u64,
+    token_identifier: TokenIdentifier,
+) {
+    let identifier_mode: NFTIdentifierMode = get_stored_value_with_user_errors::<u8>(
+        IDENTIFIER_MODE,
+        NFTCoreError::MissingIdentifierMode,
+        NFTCoreError::InvalidIdentifierMode,
+    )
+    .try_into()
+    .unwrap_or_revert();
+    let forward_uref = get_uref(
+        FORWARD_TRACKER,
+        NFTCoreError::MissingForwardTracker,
+        NFTCoreError::InvalidForwardTracker,
+    );
+    let reverse_uref = get_uref(
+        REVERSE_TRACKER,
+        NFTCoreError::MissingReverseTracker,
+        NFTCoreError::InvalidReverseTracker,
+    );
+    if storage::dictionary_get::<u64>(reverse_uref, &token_identifier.get_dictionary_item_key())
+        .unwrap_or_revert()
+        .is_some()
+    {
+        runtime::revert(NFTCoreError::InvalidTokenIdentifier)
+    }
+    match identifier_mode {
+        NFTIdentifierMode::Ordinal => {
+            if storage::dictionary_get::<u64>(
+                forward_uref,
+                &current_number_of_minted_tokens.to_string(),
+            )
+            .unwrap_or_revert()
+            .is_some()
+            {
+                runtime::revert(NFTCoreError::InvalidTokenIdentifier)
+            }
+
+            storage::dictionary_put(
+                forward_uref,
+                &current_number_of_minted_tokens.to_string(),
+                token_identifier.get_index().unwrap_or_revert(),
+            )
+        }
+        NFTIdentifierMode::Hash => {
+            if storage::dictionary_get::<String>(
+                forward_uref,
+                &current_number_of_minted_tokens.to_string(),
+            )
+            .unwrap_or_revert()
+            .is_some()
+            {
+                runtime::revert(NFTCoreError::InvalidTokenIdentifier)
+            }
+            storage::dictionary_put(
+                forward_uref,
+                &current_number_of_minted_tokens.to_string(),
+                token_identifier.clone().get_hash().unwrap_or_revert(),
+            )
+        }
+    }
+    storage::dictionary_put(
+        reverse_uref,
+        &token_identifier.get_dictionary_item_key(),
+        current_number_of_minted_tokens,
+    );
+}
+
+pub(crate) fn make_page_dictionary_item_key(
+    token_owner_key: &Key,
+    current_page_number: u64,
+) -> String {
+    let mut preimage = Vec::new();
+    preimage.append(&mut token_owner_key.clone().to_bytes().unwrap_or_revert());
+    preimage.append(&mut current_page_number.to_bytes().unwrap_or_revert());
+
+    let key_bytes = runtime::blake2b(&preimage);
+    base16::encode_lower(&key_bytes)
+}
+
+pub(crate) fn get_token_number(token_identifier: &TokenIdentifier) -> u64 {
```

Comment:
`token_index`

Reply:
in the case of TokenIdentifier::Ordinal just return the inner value

Reply:
Applied suggestion and rename